replica/table.cc

```diff
@@ -983,16 +997,26 @@ utils::chunked_vector<compaction_group*> table::compaction_groups() const {
     utils::chunked_vector<compaction_group*> ret;
     _sg_manager->for_each_storage_group([&] (size_t, storage_group& sg) {
         for (auto& cg : sg.compaction_groups()) {
-            ret.push_back(cg);
+            ret.push_back(cg.get());
         }
     });
     return ret;
 }
 
 future<> table::parallel_foreach_compaction_group(std::function<future<>(compaction_group&)> action) {
-    co_await _sg_manager->parallel_foreach_storage_group([&] (size_t, storage_group& sg) -> future<> {
+    co_await _sg_manager->parallel_foreach_storage_group([&] (storage_group& sg) -> future<> {
+        co_await utils::get_local_injector().inject("foreach_compaction_group_wait", [this, &sg] (auto& handler) -> future<> {
+            tlogger.info("foreach_compaction_group_wait: waiting");
+            while (!handler.poll_for_message() && !_async_gate.is_closed() && !sg.async_gate().is_closed()) {
+                co_await sleep(std::chrono::milliseconds(5));
+            }
+            tlogger.info("foreach_compaction_group_wait: released");
+        });
+
         for (auto& cg : sg.compaction_groups()) {
```

Comment:
We lost the inter-compaction_group parallelism here (pre-existing issue).
Why aren't we calling parallel_for_each here?

Reply:
we should. sometimes I am worried about queuing too much work, that cannot happen simulteneaously anyway. for example, only one split is allowed per shard. But we're also bound by number of tablets in a shard, so it's not like it can grow unboundly. Ok, will parallelize compaction groups of a single tablet.